### ISSUE TITLE
Exception under heavy load

### ISSUE THREAD

**Issue by SGrondin** [mirage/ocaml-dns#39]
Under heavy load, printing seems to cause this to happen:

```
retrying query for 1 times
retrying query for 3 times
Fatal error: exception (Invalid_argument Lwt.wakeup_result)
Raised at file "format.ml", line 185, characters 41-52
Called from file "format.ml", line 427, characters 6-24
```

The first 2 lines are obviously ocaml-dns.
The "Fatal error" seems to be this: https://github.com/ocsigen/lwt/blob/e995d3eefb7299f0f7209314f4d732eba82ff27f/src/core/lwt.ml#L181

The exception itself seems to be raised here: https://github.com/ocaml/ocaml/blob/14764ec8413f684359fdd677633349d842af6157/stdlib/format.ml#L185

Could be my own printing elsewhere (I only do it through Lwt_io, though), but this exception happens immediately after ocaml-dns prints its stuff and seems to be caused by it.

**Comment by dsheets** [comment]
I believe this exception was caused by incorrect handling of nameserver timeouts which are more likely to occur when the server is being hammered and drops some queries or throttles specific hosts.

I think the references to `"format.ml"` were a red herring due to Lwt exception strangeness and nearby printing code. d4be0501a42fec348a951a41386868dec4aac668 fixed an issue with very similar symptoms to this issue. That fix has now been released in opam as dns 0.15.1.

If you could please report back here, at your leisure, whether you still see this issue under heavy resolution load, we would be very appreciative.

Thanks for your report!

**Comment by SGrondin** [comment]
Free free to ping me again in a few weeks if I haven't gotten back to you on this then.

**Comment by avsm** [comment]
assuming this is now fixed or otherwise irrelevant due to #159 being merged
